src/slack/app.rs

```diff
@@ -0,0 +1,357 @@
+    user: SlackUser,
+    team: SlackIDData,
+}
+
+#[derive(Debug)]
+pub struct SlackClientHandlers {
+    im: Arc<BasicClient>,
+    identity: Arc<BasicClient>,
+}
+
+/**
+ * First redirect that handles getting authorization for identity scopes
+ */
+fn redirect_identity(client: web::Data<SlackClientHandlers>, session: Session) -> impl Responder {
+    let (authorize_url, csrf_state) = client.identity.authorize_url(CsrfToken::new_random);
+    session
+        .set("identity_csrf_state", csrf_state.secret().clone())
+        .map(|_| {
+            HttpResponse::Found()
+                .header(http::header::LOCATION, authorize_url.to_string())
+                .finish()
+        })
+}
+
+/**
+ * Second redirect that handles getting authorization for opening a channel for a direct message
+ */
+fn redirect_im(
+    client: web::Data<SlackClientHandlers>,
+    session: Session,
+    query: web::Query<Auth>,
+) -> impl Responder {
+    let (authorize_url, csrf_state) = client.im.authorize_url(CsrfToken::new_random);
+    let state = CsrfToken::new(query.state.clone());
+    if let Some(ref must_state) = session.get::<String>("identity_csrf_state").unwrap() {
+        if must_state != state.secret() {
+            return session.set("im_csrf_state", "").map(|_| {
+                HttpResponse::Found()
+                    .header(http::header::LOCATION, "/e?identityAdded=error")
+                    .finish()
+            });
+        }
+    } else {
+        return session.set("im_csrf_state", "").map(|_| {
+            HttpResponse::Found()
+                .header(http::header::LOCATION, "/e?identityAdded=error")
+                .finish()
+        });
+    }
+    session
+        .set("im_csrf_state", csrf_state.secret().clone())
+        .map(|_| {
+            HttpResponse::Found()
+                .header(http::header::LOCATION, authorize_url.to_string())
+                .finish()
+        })
+}
+
+fn auth<T: AsyncCisClientTrait + 'static>(
+    cis_client: web::Data<T>,
+    user_id: UserId,
+    query: web::Query<Auth>,
+    slack_uri_data: web::Data<SlackUriData>,
+    session: Session,
+) -> Box<dyn Future<Item = HttpResponse, Error = Error>> {
+    let state = CsrfToken::new(query.state.clone());
+    let slack_token_url = format!(
+        "{}{}&code={}",
+        TOKEN_URL,
+        slack_uri_data.slack_auth_params,
+        query.code.clone()
+    );
+
+    // Check state token from im_crsf_state
+    if let Some(ref must_state) = session.get::<String>("im_csrf_state").unwrap() {
+        if must_state != state.secret() {
+            return Box::new(future::ok(
+                HttpResponse::Found()
+                    .header(http::header::LOCATION, "/e?identityAdded=error")
+                    .finish(),
+            ));
+        }
+    } else {
+        return Box::new(future::ok(
+            HttpResponse::Found()
+                .header(http::header::LOCATION, "/e?identityAdded=error")
+                .finish(),
```

Comment:
I can count 7 instances of a `HttpResponse` of this form, I think we probably want to abstract this to a `redirect_to` function or something.

There's also a number of different error states all redirecting to the same generic error page. Even if we don't surface the difference in these errors to users, it would be nice to log them so we can debug in future if need be.

Reply:
Yea I see what you're referring to. Let me see what I can do about that.